bench/micro/Bench/Database/LSMTree/Internal/Index.hs

```diff
@@ -0,0 +1,119 @@
+import           Control.DeepSeq (rnf)
+import           Control.Monad.ST.Strict (runST)
+import           Criterion.Main (Benchmark, Benchmarkable, bench, bgroup, env,
+                     whnf)
+import           Data.List (foldl')
+import           Database.LSMTree.Extras.Generators (getKeyForIndexCompact,
+                     mkPages, toAppends)
+                     -- also for @Arbitrary@ instantiation of @SerialisedKey@
+import           Database.LSMTree.Extras.Index (Append, append)
+import           Database.LSMTree.Internal.Index (Index,
+                     IndexType (Compact, Ordinary), newWithDefaults, search,
+                     unsafeEnd)
+import           Database.LSMTree.Internal.Serialise
+                     (SerialisedKey (SerialisedKey))
+import           Test.QuickCheck (choose, vector)
+import           Test.QuickCheck.Gen (Gen (MkGen))
+import           Test.QuickCheck.Random (mkQCGen)
+
+-- * Benchmarks
+
+benchmarks :: Benchmark
+benchmarks = bgroup "Bench.Database.LSMTree.Internal.Index" $
+             map (uncurry benchmarksForSingleType)          $
+             [("Compact", Compact), ("Ordinary", Ordinary)]
+    where
+
+    benchmarksForSingleType :: String -> IndexType -> Benchmark
+    benchmarksForSingleType indexTypeName indexType
+        = bgroup (indexTypeName ++ " index") $
+          [
+              -- Search
+              env (return $ searchIndex indexType 10000) $ \ index ->
+              env (return $ searchKeys 1000)             $ \ keys  ->
+              bench "Search" $
+              searchBenchmarkable index keys,
+
+              -- Incremental construction
+              env (return $ incrementalConstructionAppends 10000) $ \ appends ->
+              bench "Incremental construction" $
+              incrementalConstructionBenchmarkable indexType appends
+          ]
+
+-- * Utilities
+
+-- | Deterministically constructs a value using a QuickCheck generator.
+generated :: Gen a -> a
+generated (MkGen exec) = exec (mkQCGen 411) 30
+
+{-|
+    Constructs serialised keys that conform to the key size constraint of
+    compact indexes.
+-}
+keysForIndexCompact :: Int             -- ^ Number of keys
+                    -> [SerialisedKey] -- ^ Constructed keys
+keysForIndexCompact = vector                                        >>>
+                      generated                                     >>>
+                      map (getKeyForIndexCompact >>> SerialisedKey)
+
+{-|
+    Constructs append operations whose serialised keys conform to the key size
+    constraint of compact indexes.
+-}
+appendsForIndexCompact :: Int      -- ^ Number of keys used in the construction
+                       -> [Append] -- ^ Constructed append operations
+appendsForIndexCompact = keysForIndexCompact                >>>
+                         mkPages 0.03 (choose (0, 16)) 0.01 >>>
+                         generated                          >>>
+                         toAppends
```

Comment:
I personally normally avoid using `QuickCheck` generators because they are often in flux, and because they are primarily tailored towards testing, not benchmarking. It can be more future proof to write these functions with `System.Random` and related functions

Reply:
I’m also not completely satisfied with using QuickCheck generators for benchmarks. That said, by using `System.Random` directly, we lose the ability to use existing utilities, like we use `mkPages` above. I’d like to leave the above code as it is for now. It might be worthwhile, though, to generally revisit the uses of QuickCheck generators in our benchmarks.

Reply:
> That said, by using `System.Random` directly, we lose the ability to use existing utilities, like we use `mkPages` above. 

That's not necessarily true. You should run the `mkPages` `Gen` with a seed, yes, but you can generate the keys with `System.Random`. See the `Index.Compact` benchmarks

Reply:
So you’re advocating a mixed-approach with some data being generated by directly using `System.Random` and other by using QuickCheck?

In the compact-index benchmark module, I can only see one use of `mkPages`, the one in `constructionEnv`. However, there the generator constructed by `mkPages` is run via `generate`, which means that it uses a seed produced by the global random number generator, not a seed specified in the source file. That said, I could still use my `generated` function for generators to work with fixed seeds.

Should I change the general index benchmarks to use that mixed approach where just the page data is generated using QuickCheck?